Implement field-sizing support for input and textarea

lukewarlow opened the pull request:
#### bcca5fc0fa1715d7da3e05cac22fb460a12539f0
<pre>
Implement field-sizing support for input and textarea
<a href="https://bugs.webkit.org/show_bug.cgi?id=269125">https://bugs.webkit.org/show_bug.cgi?id=269125</a>

Reviewed by Aditya Keerthi.

Inputs and textareas with field-sizing: content now size based on their content size not intrinsic.

This patch does not include the handling for caret width, autofill, or placeholders.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-input-number-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-input-text-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-textarea-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-textarea.html:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-input-number-expected.txt: Added.
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-input-text-expected.txt: Added.
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-textarea-expected.txt: Added.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-input-number-expected.txt: Added.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-input-text-expected.txt: Added.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-select-expected.txt: Added.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-select.tentative-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-textarea-expected.txt: Added.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-textarea.tentative-expected.txt: Removed.
* LayoutTests/platform/mac-monterey/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-input-text-expected.txt: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-input-number-expected.txt: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-input-text-expected.txt: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-textarea-expected.txt: Added.
* Source/WebCore/rendering/RenderTextControl.cpp:
(WebCore::RenderTextControl::computeLogicalHeight const):
(WebCore::RenderTextControl::computeIntrinsicLogicalWidths const):
(WebCore::RenderTextControl::computePreferredLogicalWidths):

Canonical link: <a href="https://commits.webkit.org/275649@main">https://commits.webkit.org/275649@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/978622b7b37f795bff54b6a1ae7520181fb02dd0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41279 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20293 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43657 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43843 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37371 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43586 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23364 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17623 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34554 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41853 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17219 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35556 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14795 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14949 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36558 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45172 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37466 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36876 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40615 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16094 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13201 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39002 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17713 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/35840 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9487 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17766 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17357 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on LayoutTests/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-textarea.html:
```diff
@@ -3,6 +3,10 @@
 <script src="/resources/testharnessreport.js"></script>
 <style>
+textarea {
+  font-family: monospace;
```
Remember to export this change to WPT.

This is needed because the test is reliant on font characteristics which aren't consistent across browsers.

lukewarlow replied:
https://github.com/web-platform-tests/wpt/pull/44524

lukewarlow commented on LayoutTests/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-input-number-expected.txt:
```diff
@@ -1,11 +1,11 @@
 
-FAIL number: Empty value assert_less_than: expected a number less than 148 but got 148
-FAIL number: Auto width and auto height with a long text assert_greater_than: expected a number greater than 148 but got 148
+FAIL number: Empty value assert_equals: expected 19 but got 23
```
The number tests fail more than others due to the height being different between field-sizing: fixed vs content. This seems to be due to the spin buttons being taller than the input is normally.

This is either a bug in the implementation of field-sizing (doubtful given number inputs are ultimately just textcontrols), a bug in the test, or a bug in WebKit's number input styling in general. Will investigate and try to solve in a follow up.

lukewarlow replied:
https://bugs.webkit.org/show_bug.cgi?id=269177

pxlcoder commented on Source/WebCore/rendering/RenderTextControl.cpp:
```diff
@@ -96,6 +96,9 @@ RenderBox::LogicalExtentComputedValues RenderTextControl::computeLogicalHeight(L
         return RenderBox::computeLogicalHeight(LayoutUnit(), LayoutUnit());
 
+    if (style().fieldSizing() == FieldSizing::Content)
+        return RenderBlockFlow::computeLogicalHeight(logicalHeight, logicalTop);
```
`RenderBox::computeLogicalHeight` for consistency with the rest of the code?

pxlcoder replied:
Does this work correctly if the textarea has a scrollbar?

lukewarlow replied:
I used RenderBlockFlow because it's the parent to RenderTextControl but I doubt there's any difference in it. I'll double check it doesn't break anything and change that.

Also yes I believe it works fine with scrollbars will double check though.

lukewarlow replied:
Changed to be RenderBox, I realised RenderBlockFlow and RenderBlock both don't override the function so its identical behaviour.

I've checked and the scrollbars work exactly as expected (and how CHromium behaves)

pxlcoder reviewed (commented):
This seems fine overall. Getting the `autofill` elements and `placeholder` working will be the more interesting change.

lukewarlow committed:
Implement field-sizing support for input and textarea
https://bugs.webkit.org/show_bug.cgi?id=269125

Reviewed by Aditya Keerthi.

Inputs and textareas with field-sizing: content now size based on their content size not intrinsic.

This patch does not include the handling for caret width, autofill, or placeholders.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-input-number-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-input-text-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-textarea-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-textarea.html:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-input-number-expected.txt: Added.
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-input-text-expected.txt: Added.
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-textarea-expected.txt: Added.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-input-number-expected.txt: Added.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-input-text-expected.txt: Added.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-select-expected.txt: Added.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-select.tentative-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-textarea-expected.txt: Added.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-textarea.tentative-expected.txt: Removed.
* LayoutTests/platform/mac-monterey/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-input-text-expected.txt: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-input-number-expected.txt: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-input-text-expected.txt: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/html/rendering/widgets/field-sizing-textarea-expected.txt: Added.
* Source/WebCore/rendering/RenderTextControl.cpp:
(WebCore::RenderTextControl::computeLogicalHeight const):
(WebCore::RenderTextControl::computeIntrinsicLogicalWidths const):
(WebCore::RenderTextControl::computePreferredLogicalWidths):

Canonical link: https://commits.webkit.org/275649@main
